File: src/python/core/finance_tree.py
from __future__ import annotations
from collections import deque
from datetime import date
from typing import Dict, List, Optional, Tuple

from models.category_node import CategoryNode
from models.transaction import Transaction


class FinanceTree:
    """Cây tài chính N-ary kết hợp HashMap để tối ưu tra cứu O(1)."""

    def __init__(self):
        self.root: CategoryNode = CategoryNode("Root", "Root", parent=None)
        self._node_map: Dict[str, CategoryNode] = {}
        self._initialize_default_structure()

    # =========================================================================
    # PHẦN NỘI BỘ: Đăng ký / Hủy đăng ký nút khỏi bảng băm
    # =========================================================================

    def _register_node(self, node: CategoryNode) -> None:
        path = node.get_path()
        self._node_map[path] = node
        for txn in node.transactions:
            txn.category_path = path

    def _unregister_subtree(self, node: CategoryNode) -> None:
        self._node_map.pop(node.get_path(), None)
        for child in node.children:
            self._unregister_subtree(child)

    def _re_register_subtree(self, node: CategoryNode) -> None:
        self._register_node(node)
        for child in node.children:
            self._re_register_subtree(child)

    def _initialize_default_structure(self) -> None:
        self._register_node(self.root)

        # --- Nhánh THU NHẬP ---
        income_root = CategoryNode("Thu nhập", "Income", self.root)
        self.root.children.append(income_root)
        self._register_node(income_root)
        for name in ["Lương", "Thưởng", "Đầu tư", "Thu nhập khác"]:
            node = CategoryNode(name, "Income", income_root)
            income_root.children.append(node)
            self._register_node(node)

        # --- Nhánh CHI TIÊU ---
        expense_root = CategoryNode("Chi tiêu", "Expense", self.root)
        self.root.children.append(expense_root)
        self._register_node(expense_root)

        default_expenses = {
            "Ăn uống": ["Đi chợ", "Ăn ngoài", "Cà phê"],
            "Di chuyển": ["Xăng xe", "Sửa xe", "Taxi/Grab"],
            "Nhà cửa": ["Tiền thuê", "Điện nước", "Internet"],
            "Sức khỏe": ["Thuốc", "Khám bệnh", "Thể thao"],
            "Giải trí": ["Phim ảnh", "Du lịch", "Mua sắm"],
            "Giáo dục": ["Học phí", "Sách vở"],
            "Chi tiêu khác": [],
        }

        for parent_name, child_names in default_expenses.items():
            parent_node = CategoryNode(parent_name, "Expense", expense_root)
            expense_root.children.append(parent_node)
            self._register_node(parent_node)
            for child_name in child_names:
                child_node = CategoryNode(child_name, "Expense", parent_node)
                parent_node.children.append(child_node)
                self._register_node(child_node)

    # =========================================================================
    # PHẦN 1: QUẢN LÝ NÚT CÂY
    # =========================================================================

    def insert_node(self, parent_path: str, new_name: str, category_type: str = None) -> Optional[CategoryNode]:
        parent = self.get_node_by_path(parent_path)
        if not parent:
            print(f"  [LỖI] Không tìm thấy danh mục cha: '{parent_path}'")
            return None

        actual_type = category_type or parent.category_type
        if actual_type == "Root":
            actual_type = "Expense"

        new_node = CategoryNode(new_name, actual_type, parent)
        success = parent.add_child(new_node)
        if success:
            self._register_node(new_node)
            print(f"  [OK] Đã thêm danh mục '{new_name}' vào '{parent.name}'.")
            return new_node
        return None

    def delete_node(self, node_path: str, mode: str = "CASCADE") -> bool:
        node = self.get_node_by_path(node_path)
        if not node:
            print(f"  [LỖI] Không tìm thấy danh mục: '{node_path}'")
            return False
        if node.parent is None or node.name == "Root":
            print("  [LỖI] Không thể xóa nút gốc.")
            return False

        parent = node.parent

        if mode.upper() == "CASCADE":
            self._unregister_subtree(node)
            parent.remove_child(node.name)
            print(f"  [OK] Đã xóa danh mục '{node.name}' (Cascade - và toàn bộ con cháu).")
            return True

        elif mode.upper() == "REPARENT":
            children_to_move = list(node.children)
            for child in children_to_move:
                self._unregister_subtree(child)
                child.parent = parent
                parent.children.append(child)
                self._re_register_subtree(child)

            if node.transactions:
                parent.transactions.extend(node.transactions)
                for txn in node.transactions:
                    txn.category_path = parent.get_path()
                node.transactions.clear()

            parent.remove_child(node.name)
            node.children.clear()
            self._node_map.pop(node_path, None)
            print(f"  [OK] Đã xóa danh mục '{node.name}' (Reparent - chuyển con và giao dịch lên '{parent.name}').")
            return True

        else:
            print(f"  [LỖI] Chế độ xóa không hợp lệ: '{mode}'. Phải là CASCADE hoặc REPARENT.")
            return False

    def rename_node(self, node_path: str, new_name: str) -> bool:
        node = self.get_node_by_path(node_path)
        if not node:
            print(f"  [LỖI] Không tìm thấy danh mục: '{node_path}'")
            return False
        if node.parent and node.parent.find_child_by_name(new_name):
            print(f"  [LỖI] Tên '{new_name}' đã tồn tại trong '{node.parent.name}'.")
            return False
        self._unregister_subtree(node)
        node.name = new_name
        self._re_register_subtree(node)
        print(f"  [OK] Đã đổi tên thành '{new_name}'.")
        return True

    # =========================================================================
    # PHẦN 2: THUẬT TOÁN DFS - TÍNH TỔNG TIỀN THEO NHÁNH (Đệ quy)
    # =========================================================================

    def calculate_total_dfs(self, node: CategoryNode = None) -> float:
        if node is None:
            node = self.root
        total = node.get_direct_total()
        for child in node.children:
            total += self.calculate_total_dfs(child)
        return total

    def calculate_total_by_path(self, path: str) -> float:
        node = self.get_node_by_path(path)
        if not node:
            print(f"  [LỖI] Không tìm thấy danh mục: '{path}'")
            return 0.0
        return self.calculate_total_dfs(node)

    def calculate_income_and_expense(self) -> Tuple[float, float]:
        income_node = self.get_node_by_path("Root/Thu nhập")
        expense_node = self.get_node_by_path("Root/Chi tiêu")
        total_income = self.calculate_total_dfs(income_node) if income_node else 0.0
        total_expense = self.calculate_total_dfs(expense_node) if expense_node else 0.0
        return total_income, total_expense

    # =========================================================================
    # PHẦN 3: THUẬT TOÁN TÌM KIẾM
    # =========================================================================

    def get_node_by_path(self, path: str) -> Optional[CategoryNode]:
        return self._node_map.get(path)

    def search_by_name(self, name: str) -> List[CategoryNode]:
        """BFS - tìm tất cả các nút có tên khớp."""
        results = []
        name_lower = name.strip().lower()
        queue = deque([self.root])
        while queue:
            current = queue.popleft()
            if current.name.lower() == name_lower:
                results.append(current)
            queue.extend(current.children)
        return results

    def search_by_name_dfs(self, name: str, start_node: CategoryNode = None) -> Optional[CategoryNode]:
        """DFS - tìm nút đầu tiên có tên khớp."""
        if start_node is None:
            start_node = self.root
        if start_node.name.lower() == name.strip().lower():
            return start_node
        for child in start_node.children:
            result = self.search_by_name_dfs(name, child)
            if result:
                return result
        return None

    def search_transactions_by_keyword(self, keyword: str) -> List[Transaction]:
        results = []
        keyword_lower = keyword.strip().lower()
        self._search_txn_recursive(self.root, keyword_lower, results)
        return results

    def _search_txn_recursive(self, node: CategoryNode, keyword: str, results: list) -> None:
        for txn in node.transactions:
            if keyword in txn.note.lower():
                results.append(txn)
        for child in node.children:
            self._search_txn_recursive(child, keyword, results)

    def search_transactions_by_date_range(self, start_date: date, end_date: date) -> List[Transaction]:
        results = []
        self._search_by_date_recursive(self.root, start_date, end_date, results)
        return results

    def _search_by_date_recursive(self, node: CategoryNode, start: date, end: date, results: list) -> None:
        for txn in node.transactions:
            if start <= txn.date <= end:
                results.append(txn)
        for child in node.children:
            self._search_by_date_recursive(child, start, end, results)

    # =========================================================================
    # PHẦN 4: PHÂN LOẠI GIAO DỊCH
    # =========================================================================

    def classify_and_add_transaction(self, txn: Transaction, category_path: str) -> bool:
        target_node = self.get_node_by_path(category_path)
        if not target_node:
            results = self.search_by_name(category_path)
            if results:
                target_node = results[0]
                print(f"  [THÔNG TIN] Tìm theo tên, dùng: '{target_node.get_path()}'")
            else:
                print(f"  [LỖI] Không tìm thấy danh mục: '{category_path}'")
                return False
        target_node.add_transaction(txn)
        return True

    # =========================================================================
    # PHẦN 5: DUYỆT VÀ HIỂN THỊ CÂY
    # =========================================================================

    def traverse_tree(self, node: CategoryNode = None, indent: int = 0) -> None:
        if node is None:
            node = self.root
            print("\n=======================================================")
            print("  SƠ ĐỒ CÂY DANH MỤC TÀI CHÍNH")
            print("=======================================================")

        prefix = "    " * indent
        total = self.calculate_total_dfs(node)
        txn_count = len(node.transactions)
        is_leaf_marker = " (*)" if (node.is_leaf() and txn_count > 0) else ""
        info = ""
        if txn_count > 0:
            info = f"[Tổng: {total:,.0f} VND, {txn_count} GD{is_leaf_marker}]"
        elif total > 0:
            info = f"[Tổng: {total:,.0f} VND]"

        print(f"{prefix}{'└── ' if indent > 0 else ''}📁 {node.name}  {info}")
        for child in node.children:
            self.traverse_tree(child, indent + 1)

    def get_all_transactions(self) -> List[Transaction]:
        all_txns: List[Transaction] = []
        self._collect_transactions_recursive(self.root, all_txns)
        all_txns.sort(key=lambda t: t.date, reverse=True)
        return all_txns

    def _collect_transactions_recursive(self, node: CategoryNode, result: list) -> None:
        result.extend(node.transactions)
        for child in node.children:
            self._collect_transactions_recursive(child, result)

    def get_node_count(self) -> int:
        return len(self._node_map)

    def get_all_leaf_paths(self) -> List[str]:
        paths = []
        self._get_leaf_paths_recursive(self.root, paths)
        return paths

    def _get_leaf_paths_recursive(self, node: CategoryNode, paths: list) -> None:
        if node.is_leaf():
            paths.append(node.get_path())
        for child in node.children:
            self._get_leaf_paths_recursive(child, paths)

File: src/java/core/FinanceTree.java
package core;

import java.time.LocalDate;
import java.util.*;
import models.CategoryNode;
import models.Transaction;

public class FinanceTree {
    private CategoryNode root;
    private Map<String, CategoryNode> nodeMap;

    public FinanceTree() {
        this.root = new CategoryNode("ROOT", "ROOT", null);
        this.nodeMap = new HashMap<>();
        initializeDefaultStructure();
    }

    public CategoryNode getRoot() {
        return root;
    }

    private void initializeDefaultStructure() {
        registerNode(root);

        // --- Nhánh THU ---
        CategoryNode thuNode = addNodeToParent("THU", "THU", root);

        addNodeToParent("Lương", "THU", thuNode);
        addNodeToParent("Thưởng", "THU", thuNode);
        addNodeToParent("Đầu tư", "THU", thuNode);
        addNodeToParent("Thu nhập khác", "THU", thuNode);

        // --- Nhánh CHI ---
        CategoryNode chiNode = addNodeToParent("CHI", "CHI", root);

        CategoryNode nhuCau = addNodeToParent("Nhu cầu thiết yếu", "CHI", chiNode);

        CategoryNode anUong = addNodeToParent("Ăn uống", "CHI", nhuCau);
        addNodeToParent("Ăn sáng", "CHI", anUong);
        addNodeToParent("Ăn trưa", "CHI", anUong);
        addNodeToParent("Ăn tối", "CHI", anUong);

        CategoryNode nhaO = addNodeToParent("Nhà ở", "CHI", nhuCau);
        addNodeToParent("Tiền thuê", "CHI", nhaO);
        addNodeToParent("Điện nước", "CHI", nhaO);

        CategoryNode diChuyen = addNodeToParent("Di chuyển", "CHI", nhuCau);
        addNodeToParent("Xăng xe", "CHI", diChuyen);
        addNodeToParent("Giao thông công cộng", "CHI", diChuyen);

        CategoryNode giaoDoc = addNodeToParent("Giáo dục & Phát triển", "CHI", chiNode);
        addNodeToParent("Sách vở", "CHI", giaoDoc);
        addNodeToParent("Khóa học", "CHI", giaoDoc);

        CategoryNode huongThu = addNodeToParent("Hưởng thụ", "CHI", chiNode);
        addNodeToParent("Du lịch", "CHI", huongThu);
        addNodeToParent("Giải trí", "CHI", huongThu);
    }

    /** Hàm tiện ích: tạo nút, gán vào cha, đăng ký vào nodeMap. */
    private CategoryNode addNodeToParent(String name, String type, CategoryNode parent) {
        CategoryNode node = new CategoryNode(name, type, parent);
        parent.getChildren().add(node);
        registerNode(node);
        return node;
    }

    public void registerNode(CategoryNode node) {
        String path = node.getPath();
        nodeMap.put(path, node);
        for (Transaction txn : node.getTransactions()) {
            txn.setCategoryPath(path);
        }
    }

    public void unregisterSubtree(CategoryNode node) {
        String path = node.getPath();
        nodeMap.remove(path);
        for (CategoryNode child : node.getChildren()) {
            unregisterSubtree(child);
        }
    }

    public void reRegisterSubtree(CategoryNode node) {
        registerNode(node);
        for (CategoryNode child : node.getChildren()) {
            reRegisterSubtree(child);
        }
    }

    // =========================================================================
    // PHẦN 1: QUẢN LÝ NÚT CÂY (Tree Node Management)
    // =========================================================================

    public CategoryNode insertNode(String parentPath, String newName, String categoryType) {
        CategoryNode parent = getNodeByPath(parentPath);
        if (parent == null) {
            return null;
        }

        String actualType = categoryType;
        if (actualType == null || actualType.isEmpty()) {
            actualType = parent.getCategoryType();
            if (actualType.equals("ROOT")) {
                actualType = "CHI"; // Mặc định là CHI khi thêm vào gốc
            }
        }

        CategoryNode newNode = new CategoryNode(newName, actualType, parent);
        boolean success = parent.addChild(newNode);

        if (success) {
            registerNode(newNode);
            return newNode;
        }
        return null;
    }

    public boolean deleteNode(String nodePath, String mode) {
        CategoryNode node = getNodeByPath(nodePath);
        if (node == null) {
            return false;
        }

        if (node.getParent() == null || node.getName().equals("ROOT")) {
            // Không cho xóa nút gốc ROOT
            return false;
        }

        CategoryNode parent = node.getParent();

        if (mode.equalsIgnoreCase("CASCADE")) {
            // Xóa toàn bộ cây con khỏi bảng băm
            unregisterSubtree(node);
            // Xóa nút khỏi danh sách con của cha
            parent.removeChild(node.getName());
            return true;
        } else if (mode.equalsIgnoreCase("REPARENT")) {
            // Lưu danh sách con tạm thời
            List<CategoryNode> childrenToMove = new ArrayList<>(node.getChildren());
            for (CategoryNode child : childrenToMove) {
                // Hủy đăng ký cây con của child khỏi bảng băm dưới đường dẫn cũ
                unregisterSubtree(child);
                
                // Gán parent mới
                child.setParent(parent);
                parent.getChildren().add(child);
                
                // Đăng ký lại cây con của child dưới đường dẫn mới
                reRegisterSubtree(child);
            }
            
            // Chuyển giao dịch trực tiếp của nút bị xóa sang parent
            if (!node.getTransactions().isEmpty()) {
                parent.getTransactions().addAll(node.getTransactions());
                for (Transaction txn : node.getTransactions()) {
                    txn.setCategoryPath(parent.getPath());
                }
                node.getTransactions().clear();
            }
            
            // Xóa nút bị xóa khỏi children của parent
            parent.removeChild(node.getName());
            node.getChildren().clear();
            
            // Xóa nút bị xóa khỏi bảng băm
            nodeMap.remove(nodePath);
            
            return true;
        } else {
            return false;
        }
    }

    public boolean renameNode(String nodePath, String newName) {
        CategoryNode node = getNodeByPath(nodePath);
        if (node == null) {
            return false;
        }

        if (node.getParent() != null && node.getParent().findChildByName(newName) != null) {
            return false;
        }

        // Xóa các key cũ trong bảng băm
        unregisterSubtree(node);
        // Đổi tên
        node.setName(newName);
        // Đăng ký lại với key mới
        reRegisterSubtree(node);
        return true;
    }

    // =========================================================================
    // PHẦN 2: THUẬT TOÁN DFS - TÍNH TỔNG TIỀN THEO NHÁNH
    // =========================================================================

    public double calculateTotalDfs(CategoryNode node) {
        if (node == null) {
            node = this.root;
        }

        // Bước 1: Tính tổng giao dịch TRỰC TIẾP tại nút này
        double total = node.getDirectTotal();

        // Bước 2: ĐỆ QUY - Cộng thêm tổng từ các nút con
        for (CategoryNode child : node.getChildren()) {
            total += calculateTotalDfs(child);
        }

        return total;
    }

    public double calculateTotalByPath(String path) {
        CategoryNode node = getNodeByPath(path);
        if (node == null) {
            return 0.0;
        }
        return calculateTotalDfs(node);
    }

    public double[] calculateIncomeAndExpense() {
        // Sau khi fix getPath(), THU và CHI được lưu trong nodeMap với key "THU" và "CHI"
        CategoryNode incomeNode = getNodeByPath("THU");
        CategoryNode expenseNode = getNodeByPath("CHI");

        double totalIncome = incomeNode != null ? calculateTotalDfs(incomeNode) : 0.0;
        double totalExpense = expenseNode != null ? calculateTotalDfs(expenseNode) : 0.0;
        return new double[]{totalIncome, totalExpense};
    }

    // =========================================================================
    // PHẦN 3: THUẬT TOÁN TÌM KIẾM (Search Algorithms)
    // =========================================================================

    public CategoryNode getNodeByPath(String path) {
        return nodeMap.get(path);
    }

    public List<CategoryNode> searchByName(String name) {
        List<CategoryNode> results = new ArrayList<>();
        String nameLower = name.trim().toLowerCase();

        Queue<CategoryNode> queue = new LinkedList<>();
        queue.add(this.root);
        while (!queue.isEmpty()) {
            CategoryNode current = queue.poll();
            if (current.getName().toLowerCase().equals(nameLower)) {
                results.add(current);
            }
            queue.addAll(current.getChildren());
        }

        return results;
    }

    public CategoryNode searchByNameDfs(String name, CategoryNode startNode) {
        if (startNode == null) {
            startNode = this.root;
        }

        if (startNode.getName().equalsIgnoreCase(name.trim())) {
            return startNode;
        }

        for (CategoryNode child : startNode.getChildren()) {
            CategoryNode result = searchByNameDfs(name, child);
            if (result != null) {
                return result;
            }
        }

        return null;
    }

    public List<Transaction> searchTransactionsByKeyword(String keyword) {
        List<Transaction> results = new ArrayList<>();
        String keywordLower = keyword.trim().toLowerCase();
        searchTxnRecursive(this.root, keywordLower, results);
        return results;
    }

    private void searchTxnRecursive(CategoryNode node, String keyword, List<Transaction> results) {
        for (Transaction txn : node.getTransactions()) {
            if (txn.getNote().toLowerCase().contains(keyword)) {
                results.add(txn);
            }
        }
        for (CategoryNode child : node.getChildren()) {
            searchTxnRecursive(child, keyword, results);
        }
    }

    public List<Transaction> searchTransactionsByDateRange(LocalDate startDate, LocalDate endDate) {
        List<Transaction> results = new ArrayList<>();
        searchByDateRecursive(this.root, startDate, endDate, results);
        return results;
    }

    private void searchByDateRecursive(CategoryNode node, LocalDate startDate, LocalDate endDate, List<Transaction> results) {
        for (Transaction txn : node.getTransactions()) {
            LocalDate d = txn.getDate();
            if ((d.isAfter(startDate) || d.isEqual(startDate)) && (d.isBefore(endDate) || d.isEqual(endDate))) {
                results.add(txn);
            }
        }
        for (CategoryNode child : node.getChildren()) {
            searchByDateRecursive(child, startDate, endDate, results);
        }
    }

    // =========================================================================
    // PHẦN 4: THUẬT TOÁN PHÂN LOẠI GIAO DỊCH (Transaction Classification)
    // =========================================================================

    public boolean classifyAndAddTransaction(Transaction txn, String categoryPath) {
        CategoryNode targetNode = getNodeByPath(categoryPath);

        if (targetNode == null) {
            // Thử tìm theo tên nếu không tìm được theo đường dẫn đầy đủ
            List<CategoryNode> results = searchByName(categoryPath);
            if (!results.isEmpty()) {
                targetNode = results.get(0);
            } else {
                return false;
            }
        }

        targetNode.addTransaction(txn);
        return true;
    }

    // =========================================================================
    // PHẦN 5: DUYỆT VÀ HIỂN THỊ CÂY (Tree Traversal & Display)
    // =========================================================================

    public String traverseTree(CategoryNode node, int indent) {
        StringBuilder sb = new StringBuilder();
        buildTreeVisualized(node, indent, sb);
        return sb.toString();
    }

    private void buildTreeVisualized(CategoryNode node, int indent, StringBuilder sb) {
        if (node == null) {
            node = this.root;
            sb.append("\n=======================================================\n");
            sb.append("  SƠ ĐỒ CÂY DANH MỤC TÀI CHÍNH\n");
            sb.append("=======================================================\n");
        }

        StringBuilder prefix = new StringBuilder();
        for (int i = 0; i < indent; i++) {
            prefix.append("    ");
        }

        double total = calculateTotalDfs(node);
        int txnCount = node.getTransactions().size();
        String isLeafMarker = (node.isLeaf() && txnCount > 0) ? " (*)" : "";

        String info = "";
        if (txnCount > 0) {
            info = String.format("[Tổng: %,.0f VND, %d GD%s]", total, txnCount, isLeafMarker);
        } else if (total > 0) {
            info = String.format("[Tổng: %,.0f VND]", total);
        }

        sb.append(prefix.toString()).append(indent > 0 ? "└── " : "").append("📁 ").append(node.getName()).append("  ").append(info).append("\n");

        for (CategoryNode child : node.getChildren()) {
            buildTreeVisualized(child, indent + 1, sb);
        }
    }

    public List<Transaction> getAllTransactions() {
        List<Transaction> allTxns = new ArrayList<>();
        collectTransactionsRecursive(this.root, allTxns);
        // Sắp xếp theo ngày mới nhất (giảm dần)
        allTxns.sort((t1, t2) -> t2.getDate().compareTo(t1.getDate()));
        return allTxns;
    }

    private void collectTransactionsRecursive(CategoryNode node, List<Transaction> result) {
        result.addAll(node.getTransactions());
        for (CategoryNode child : node.getChildren()) {
            collectTransactionsRecursive(child, result);
        }
    }

    public int getNodeCount() {
        return nodeMap.size();
    }

    public List<String> getAllLeafPaths() {
        List<String> paths = new ArrayList<>();
        getLeafPathsRecursive(this.root, paths);
        return paths;
    }

    private void getLeafPathsRecursive(CategoryNode node, List<String> paths) {
        if (node.isLeaf()) {
            paths.add(node.getPath());
        }
        for (CategoryNode child : node.getChildren()) {
            getLeafPathsRecursive(child, paths);
        }
    }
}
